### PR TITLE
Joost world parser

### PR DESCRIPTION
**Pull request by joostvanrijsinge**
World parser ported and working with assetsmanager

### CODE REVIEW

File: res/test3.tmx [hunk 1]
```diff
@@ -0,0 +1,113 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jorgdebont) [inline res/test3.tmx]: naming convention ?
Test1
Test2
test3 

i mean it doesnt really matter

File: src/gamelogic/Level.cpp [hunk 2]
```diff
@@ -36,7 +37,7 @@ namespace GameLogic {
 
     TileObject* Level::get_tile(int x, int y)
     {
-        return _field.at(x).at(y); //TODO: use at or []?
+        return _field[x][y];
```
Comment (jorgdebont) [inline src/gamelogic/Level.cpp]: What was the decision to actually change this?

File: src/gamelogic/Level.hpp [hunk 3]
```diff
@@ -38,7 +38,7 @@ namespace GameLogic {
          */
 
     public:
-        Level(vector<vector<Tile*>> field, Engine::SPTR_AssetsManager assets);
+        Level(vector<vector<Tile*>> field, CoordinateDouble spawnpoint, Engine::SPTR_AssetsManager assets);
```
Comment (jorgdebont) [inline src/gamelogic/Level.hpp]: is it actually a double or an int?

File: src/gamelogic/RapidXMLAdapter.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    /// \brief gets the objects from loaded .tmx
+    /// 
+    /// This function returns the objects that is used in this .tmx.
+    /// RapidXMLAdapter::setup_document has to be called before using this 
+    vector<tuple<size_t, size_t, char*>> RapidXMLAdapter::get_objects()
+    {
+        vector<tuple<size_t, size_t, char*>> object_list;
+        for ( xml_node<> * object_node = _object_group_node->first_node( "object" ); object_node; object_node = object_node->next_sibling() ) {
+            size_t x = std::stoi( object_node->first_attribute( "x" )->value() ) / 64;
```
Comment (jorgdebont) [inline src/gamelogic/RapidXMLAdapter.cpp]: 64 is a magic number

File: src/gamelogic/RapidXMLAdapter.hpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#include <string>
+#include <vector>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <cstring>
+#include <tuple>
+#include "../../libs/RapidXML/rapidxml.hpp"
+
+namespace GameLogic {
```
Comment (jorgdebont) [inline src/gamelogic/RapidXMLAdapter.hpp]: is it actually gamelogic?

File: src/gamelogic/RapidXMLAdapter.hpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        string get_texture_source();
+        size_t get_tile_height();
+        size_t get_tile_width();
+        size_t get_tile_count();
+        size_t get_tile_column_size();
+        vector<vector<size_t>> get_map();
+        vector<tuple<size_t, size_t, char*>> get_objects();
+    private:
+        vector<char> _buffer;
+        xml_document<> * _doc;
```
Comment (jorgdebont) [inline src/gamelogic/RapidXMLAdapter.hpp]: You should really change you editor settings to the styleguide
please remove the space between <> and *

File: src/gamelogic/WorldParser.cpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
+#include "WorldParser.hpp"
+#include <tuple>
+
+namespace GameLogic {
+
+    using std::tuple;
+    using std::get;
+
+    WorldParser::WorldParser()
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: empty constructor and deconstructor, i dont care but the cpp "experts" care

File: src/gamelogic/WorldParser.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        for ( size_t i = 0; i < object_list.size(); i++ ) {
+            if ( std::strcmp( get<2>( object_list[i] ), "PlayerSpawn" ) == 0 ) {
+                if ( map[get<1>( object_list[i] )][get<0>( object_list[i] )]->is_wall() ) {
+                    throw exception( "file invalid spawn point inside wall" );
+                }
+                else {
+                    spawn_point = CoordinateDouble { get<1>( object_list[i] ) + 0.5, get<0>( object_list[i] ) + 0.5 };
+                }
+            }
+            // Other Objects
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: commented code

File: src/gamelogic/WorldParser.cpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    /// \brief Generates a level from the source .tmx file
+    /// 
+    /// This function generates a Level object which has:
+    /// * 2D vector of Tiles
+    /// * TileSet struct
+    /// needs to be inside try catch
+    ///
+    /// \param file_location location of the file to be used
+    /// \return returns a Level
+    Level WorldParser::generate_level( std::string file_location, Engine::SPTR_AssetsManager assets )
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: this bby is a bit long, split it

File: src/gamelogic/WorldParser.cpp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        CoordinateDouble spawn_point;
+        rapid_adapter->setup_document( file_location );
+
+        if ( !assets->init( rapid_adapter->get_texture_source(), rapid_adapter->get_tile_width(), rapid_adapter->get_tile_height(), rapid_adapter->get_tile_count() ) ) {
+            std::cout << "AssetsManager has not initted correctly." << std::endl;
+        }
+
+        int_map = rapid_adapter->get_map();
+
+        for ( int y = 1; y < int_map.size(); y++ ) {
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: why start at y = 1?

File: src/gamelogic/WorldParser.cpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            vector<Tile*> map_row;
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: this is a copy, you can just pushback here already

File: src/gamelogic/WorldParser.cpp [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            }
+
+            map.push_back( map_row );
+        }
+
+        object_list = rapid_adapter->get_objects();
+
+        for ( size_t i = 0; i < object_list.size(); i++ ) {
+            if ( std::strcmp( get<2>( object_list[i] ), "PlayerSpawn" ) == 0 ) {
+                if ( map[get<1>( object_list[i] )][get<0>( object_list[i] )]->is_wall() ) {
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: this is a big cluster fuck of variables, try to make it a bit more readable
perhaps by int x = ... and y = ...
map[x][y] or whatever order

Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: if you check for a wall, check also if it exist on our map

File: src/gamelogic/WorldParser.cpp [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                    throw exception( "file invalid spawn point inside wall" );
+                }
+                else {
+                    spawn_point = CoordinateDouble { get<1>( object_list[i] ) + 0.5, get<0>( object_list[i] ) + 0.5 };
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: same story

Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: magic number 0.5

File: src/gamelogic/WorldParser.cpp [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+                }
+            }
+            // Other Objects
+            //else if ( std::strcmp(get<2>( object_list[i] ), "Other_Object" ) == 0 ) {
+            //}
+        }
+
+
+        // generated_level needs to be deleted in the mainclass/gameloop when this level has been completed/finished/player quits.
+        Level generated_level = Level( map, spawn_point, assets );
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: it currently isnt a pointer, so no need for comment on deletion

File: src/gamelogic/WorldParser.cpp [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+        return generated_level;
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: just return it in 1 step instead of putting it in a var

File: src/gamelogic/WorldParser.hpp [hunk 16]
```diff
@@ -0,0 +1,26 @@
+#ifndef WorldParser_HPP
+#define WorldParser_HPP
+
+#include <string>
+#include <vector>
+#include <cstring>
+//#include "TileSet.hpp"
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.hpp]: commented code

File: res/test3.tmx (hunk 1 above)
Comment (joostvanrijsinge) [inline res/test3.tmx]: True

File: src/gamelogic/Level.cpp (hunk 2 above)
Comment (joostvanrijsinge) [inline src/gamelogic/Level.cpp]: https://stackoverflow.com/questions/6911289/should-i-use-stdvectorat-in-my-code

Comment (joostvanrijsinge) [inline src/gamelogic/Level.cpp]: deze had jij naar mij gestuurd eerder

File: src/gamelogic/RapidXMLAdapter.hpp (hunk 5 above)
Comment (joostvanrijsinge) [inline src/gamelogic/RapidXMLAdapter.hpp]: well we put it as not being engine and I'm not sure what else to make it then

File: src/gamelogic/WorldParser.cpp (hunk 10 above)
Comment (joostvanrijsinge) [inline src/gamelogic/WorldParser.cpp]: this is because the Tiled map starts with an empty row

File: src/gamelogic/WorldParser.cpp (hunk 11 above)
Comment (joostvanrijsinge) [inline src/gamelogic/WorldParser.cpp]: I don't know what you mean here

File: src/gamelogic/WorldParser.cpp [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    /// 
+    /// This function generates a Level object which has:
+    /// * 2D vector of Tiles
+    /// * TileSet struct
+    /// needs to be inside try catch
+    ///
+    /// \param file_location location of the file to be used
+    /// \param assets the assets manager for this .tmx
+    /// \return returns a Level
+    Level WorldParser::generate_level( std::string file_location, Engine::SPTR_AssetsManager assets )
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: This function is still far too long

File: src/gamelogic/Game.cpp [hunk 18]
```diff
@@ -18,20 +20,23 @@ namespace GameLogic {
     {
         this->SDL_facade.init();
         this->init_states();
-        WorldParser parser;
-        Engine::SPTR_AssetsManager assets = std::make_shared<AssetsManager>(this->SDL_facade);
+        //WorldParser parser;
```
Comment (jorgdebont) [inline src/gamelogic/Game.cpp]: why would you comment out code?

File: src/gamelogic/Game.cpp [hunk 19]
```diff
@@ -65,6 +70,33 @@ namespace GameLogic {
         this->_new_state = state;
     }
 
+    bool Game::load_Level()
+    {
+        WorldParser parser;
+        Engine::SPTR_AssetsManager assets = std::make_shared<AssetsManager>( this->SDL_facade );
+
+        try {
+            this->_level = { std::make_shared<Level>( parser.generate_level( VIDYA_RUNPATH + "res/Test2.tmx", assets ) ) };
```
Comment (jorgdebont) [inline src/gamelogic/Game.cpp]: this is fucking hardcoded in the method

File: src/gamelogic/Game.cpp [hunk 20]
```diff
@@ -65,6 +70,33 @@ namespace GameLogic {
...
+        }
+        catch ( const Exceptions::FileInvalidException& e ) {
+            std::cout << e.what() << std::endl;
+            std::cout << "Returning to Menu" << std::endl;
+            return false;
+        }
+        catch ( const std::exception& e ) {
+            std::cout << e.what();
```
Comment (jorgdebont) [inline src/gamelogic/Game.cpp]: why no std::endl here if you use it above

File: src/gamelogic/Game.hpp [hunk 21]
```diff
@@ -33,6 +33,7 @@ namespace GameLogic {
 
         void init_states();
         void set_new_state(SPTR_IGameState state);
+        bool load_Level(); //TODO: load level based on file location
```
Comment (jorgdebont) [inline src/gamelogic/Game.hpp]: no not a todo, do this now

File: src/gamelogic/WorldParser.cpp [hunk 22]
```diff
@@ -35,7 +35,7 @@ namespace GameLogic {
         vector<tuple<size_t, size_t, char*>> object_list;
         CoordinateDouble spawn_point;
         rapid_adapter->setup_document( file_location );
-        string path = "res\\" + rapid_adapter->get_texture_source();
+        string path = file_location.substr( 0, file_location.find_last_of( "\\/" ) ) + "/" + rapid_adapter->get_texture_source();
```
Comment (jorgdebont) [inline src/gamelogic/WorldParser.cpp]: what?! why how, why?

File: src/gamelogic/state/LoadState.cpp [hunk 23]
```diff
@@ -19,7 +20,7 @@ namespace State {
         this->_context.SDL_facade.clear_screen();
         this->_context.SDL_facade.draw_rect({0,0}, 640, 480, Color{0,0,0});
 
-        this->_loading_screen_time = this->_loading_screen_time + 2; //TODO: To ticks
```
Comment (jorgdebont) [inline src/gamelogic/state/LoadState.cpp]: what?

Comment (jorgdebont) [inline src/gamelogic/state/LoadState.cpp]: why would you add debug code, plz dont use git add -A but use git add -p, so you can actually review your changes

**Comment by ThomasvLingen** [comment]
We fix

**Pull request merged**